### ISSUE TITLE
Godot 4.0 stable - autocomplete for parent classes' properties and methods not working

### ISSUE THREAD

**Issue by zeroxx1986** [godotengine/godot#75040]
### Godot version

4.0 stable (92bee43ad)

### System information

Windows 10 x64 22H2

### Issue description

GDscript autocomplete for parent classes' inherited properties does not work. Not sure if it's a global problem, or just specific to the one I've found (completely new to Godot, actually was just following through the 2D tutorial).

### Steps to reproduce

1. Create a `Path2D` (named MobPath) and then a `PathFollow2D` as a child (named MobSpawnLocation)
2. Refer to it in code either via `$MobPath/MobSpawnLocation` or `get_node("MobPath/MobSpawnLocation")` (doesn't matter)
3. Try to find the `position` or `rotation` properties of this object - autocomplete does not offer it up.

Both of these properties are inherited from the `Node2D` parent class.

Tried finding any other inherited property or class coming from `Node2D`, not working.
Tried finding any other inherited property or class coming from other parent classes, not working either.

Example: part of the script from the tutorial

```gdscript
func _on_mob_timer_timeout():
	var mob = mob_scene.instantiate()
	var mob_spawn_location = $MobPath/MobSpawnLocation
	mob_spawn_location.progress_ratio = randf()
	
	var direction = mob_spawn_location.rotation + PI/2    # .rotation is not found via autocomplete
	mob.position = mob_spawn_location.position            # .position is not found at all via autocomplete
	
	direction += randf_range(-PI / 4, PI / 4)
	mob.rotation = direction
	
	var velocity = Vector2(randf_range(150.0, 250.0), 0.0)
	mob.linear_velocity = velocity.rotated(direction)
	
	add_child(mob)
```

### Minimal reproduction project

N/A

**Comment by Calinou** [comment]
> Both of these properties are inherited from the Node2D parent class.

The result of `$...` is a Node, which does not necessarily have a `position` or `rotation` property.

**Comment by zeroxx1986** [comment]
I'm not sure I follow you.. The `MobSpawnLocation` object is an instance of the `PathFollow2D` class, and according to [Godot Docs](https://docs.godotengine.org/en/stable/classes/class_pathfollow2d.html), this "**Inherits**: [Node2D](https://docs.godotengine.org/en/stable/classes/class_node2d.html#class-node2d) < [CanvasItem](https://docs.godotengine.org/en/stable/classes/class_canvasitem.html#class-canvasitem) < [Node](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node) < [Object](https://docs.godotengine.org/en/stable/classes/class_object.html#class-object)".
Shouldn't inheritance work in a way that anything which is a public property of any of your classes' parents, is by default an accessible property of your class, unless it's specifically made private in your class? But then if it was made private, then accessing those properties should not work at all - but it does work.

**Comment by AThousandShips** [comment]
The function `get_node` returns a `Node` so it doesn't have hints for things like `Node2D`, this has nothing to do with hidden properties, it's all to do with that the engine doesn't know what type that node is, because it can change

**Comment by zeroxx1986** [comment]
Fair enough - I've added the type hint into the editor, yet autocomplete is not working:

![image](https://user-images.githubusercontent.com/4190724/226754995-e2423e95-bdbc-408e-a56b-9c86f73b5108.png)

**Comment by zeroxx1986** [comment]
Even tried it with Node2D as type hint - not working:

![image](https://user-images.githubusercontent.com/4190724/226755310-bf05dd65-760d-4754-b851-e4fd943af4b0.png)

**Comment by Calinou** [comment]
Does it work if you use `var mob_spawn_location := $MobPath/MobSpawnLocation as PathFollow2D`?

**Comment by zeroxx1986** [comment]
Yes, it does!

![image](https://user-images.githubusercontent.com/4190724/226763686-a2901415-7f82-487a-b492-76a618449d67.png)

**Comment by KoBeWi** [comment]
Can't reproduce™
![image](https://user-images.githubusercontent.com/2223172/226767536-03d211ce-456d-4abf-bf22-b7ce959665e5.png)
![image](https://user-images.githubusercontent.com/2223172/226767546-b333fc90-0032-4ba1-a585-e86541c6623b.png)

**Comment by zeroxx1986** [comment]
I believe the scene is opened, although I'm not sure what this would mean exactly. You can see to the left under Scene, that Main is selected, MobPath/SpawnLocation is also visible. I've rewritten the code to be the first variant but with type hinting, where the autocomplete does not work.

![image](https://user-images.githubusercontent.com/4190724/226770731-c301df75-37f6-4d56-b7d1-f6a763dc8613.png)

**Comment by ajreckof** [comment]
I've tested and can reproduce. Here is a minimal reproduction code 
```gdscript
func _ready():
	var child = $Node2D
	rotation = child.rotation # rotation won't autocomplete here
```
To reproduce you need both to store the node in a variable and to try to autocomplete in an assignement


[MRP.zip](https://github.com/godotengine/godot/files/11346280/MRP.zip)

**Comment by MBoffin** [comment]
I have noticed similar behavior with autocomplete not working, even with type hints, though having nothing to do with parent classes (as noted in the title). The following GIF is from version v4.1.2.stable.official [399c9dc39] and the `level` variable is a `PackedScene` `@export` variable.

As far as I understand it, the type hint of `: TileMap` should be making `tilemap` into type `TileMap` with appropriate autocomplete. However, only adding `as TileMap` gets the autocomplete working.

![autocomplete](https://github.com/godotengine/godot/assets/673691/b60f0080-4616-485d-a500-37a4e0cb4eda)

**Comment by HolonProduction** [comment]
For the problem with type hints not working see #74888.

The unique problem here is what was described by ajreckof.

**Comment by publicJorn** [comment]
I'm new using godot and following the 2d tutorial. If it helps: I'm running into the same thing here https://docs.godotengine.org/en/latest/getting_started/first_2d_game/05.the_main_game_scene.html when writing the `_on_mob_timer_timeout` function.

context:
- I have set Editor settings - Text Editor - Completion - Add type hints = On.

main.gd  
```
extends Node

@export var mob_scene: PackedScene

func _on_mob_timer_timeout() -> void:
    var mob = mob_scene.instantiate()
    # ...
```

mob_scene is set to the Mob scene, which is a RigidBody2D which is a descendant of Node2D amongst others.  
But `mob.position` or `mob.rotation` do not autocomplete.

When I add a class definition to `mob.gd` like so: `class_name Mob`, I can type cast with `as Mob` and then it works.  
But it doesn't make much sens to me to type cast, since it's not a type-safe language to start with.

It would be really nice if it would just infer all ancestor members automatically.

**Comment by HolonProduction** [comment]
> I've tested and can reproduce. Here is a minimal reproduction code
> 
> func _ready():
> 	var child = $Node2D
> 	rotation = child.rotation # rotation won't autocomplete here
> 
> To reproduce you need both to store the node in a variable and to try to autocomplete in an assignement
> 
> [MRP.zip](https://github.com/godotengine/godot/files/11346280/MRP.zip)

Can't reproduce anymore
